autopush/websocket.py

```diff
@@ -352,63 +346,40 @@ def process_hello(self, data):
         # This must be a helo, or we kick the client
         cmd = data.get("messageType")
         if cmd != "hello":
-            self.sendClose()
-            return
+            return self.sendClose()
 
         if self.uaid:
-            self.returnError("hello", "duplicate hello", 401)
-            return
+            return self.returnError("hello", "duplicate hello", 401)
 
         uaid = data.get("uaid")
-        valid = False
-        if uaid:
-            try:
-                valid = bool(uuid.UUID(uaid))
-            except ValueError:
-                pass
-        if not uaid or not valid:
-            uaid = str(uuid.uuid4())
+        _, uaid = validate_uaid(uaid)
         self.uaid = uaid
 
-        connect = data.get("connect")
-        if connect and self.ap_settings.bridge:
-            self.transport.pauseProducing()
-            d = self.deferToThread(self.ap_settings.bridge.register, uaid,
-                                   connect)
-            d.addCallback(self._check_router, True)
-            d.addErrback(self.err_hello)
-        else:
-            self._check_router(False)
+        # Default router choice
+        router_type = data.get("router_type", "simplepush")
+        if router_type not in self.ap_settings.routers:
+            return self.returnError("hello", "invalid router", 401)
 
-    def _check_router(self, paused=False, bridge_register=None):
-        if paused:
-            self.transport.resumeProducing()
-        # Bridge registration either succeeds and returns true,
-        # is not specified by the hello, or fails with an exception.
-        router = self.ap_settings.router
-        url = self.ap_settings.router_url
         self.transport.pauseProducing()
-        d = self.deferToThread(router.register_user,
-                               self.uaid, url, self.connected_at)
+        user_item = dict(
+            uaid=self.uaid,
+            node_id=self.ap_settings.router_url,
+            connected_at=self.connected_at,
+            router_type=router_type,
+            router_data={},
```

Comment:
Correct, RESTful HTTP API is the *only* way to register external routers or anything other than simplepush, and webpush.